### DISCUSSION TITLE
How to build libvips with openslide support

### DISCUSSION THREAD

**Discussion by mgbee8** [libvips/libvips#4021] [Q&A]
We are using NetVips.native to process whole slide images. We using docker and kubernetes to deploy.
We are using the imagemcr.microsoft.com/dotnet/runtime:8.0-jammy
When we try to perform openslide operations using this configuration we have this error 
no such operation openslideload

We then attempted to build libvips with openslide support

an abridged version of the docker file is here

```
RUN apt update && apt-get install -o Acquire::https::Verify-Peer=false -y --no-install-recommends \
	build-essential \
	cmake \
	meson \
	unzip \
	autoconf \
	automake \
	libtool \
	bc \
	zlib1g-dev \
	libxml2-dev \
	libjpeg8-dev \
	libopenjp2-7-dev \
	libtiff-dev \
	libglib2.0-dev \
	libgdk-pixbuf-2.0-dev \
	libsqlite3-dev \
	sqlite3 \
	libvips-dev \
	libexif-dev \
	liblcms2-dev \
	fftw-dev \
	libgif-dev \
	libpng-dev \
	libwebp-dev \
	liborc-0.4-dev \
	libpoppler-dev \
	librsvg2-dev \
	libgsf-1-dev \
	libopenexr-dev \
	gtk-doc-tools \
	libjpeg-turbo8 \
	wget

WORKDIR /usr/local/src

ARG OPENSLIDE_VERSION=4.0.0
ARG OPENSLIDE_URL=https://github.com/openslide/openslide/releases

# OpenSlide
# https://github.com/openslide/openslide
WORKDIR /tmp/
RUN wget https://github.com/openslide/openslide/archive/refs/tags/v4.0.0.zip 
RUN unzip -qq v4.0.0.zip
WORKDIR /tmp/openslide-4.0.0
RUN meson setup --libdir=lib --buildtype=release builddir && meson compile -C builddir && meson install -C builddir

#RUN export PKG_CONFIG_PATH=/usr/local/vips  
#RUN	export PATH=/usr/local/vips/bin:$PATH
#RUN chmod +x configure
#RUN ./configure --prefix=/usr/local/vips

RUN export LD_LIBRARY_PATH=/tmp/openslide-4.0.0/lib \
    && export PATH=/tmp/openslide-4.0.0:$PATH \
    && export PKG_CONFIG_PATH=/tmp/openslide-4.0.0

ARG VIPS_VERSION=8.15.2
ARG VIPS_URL=https://github.com/libvips/libvips/releases/download/

# Vips
# https://github.com/libvips/libvips/wiki/Build-for-Ubuntu
WORKDIR /tmp/
RUN wget https://github.com/libvips/libvips/archive/refs/tags/v8.15.2.zip
RUN unzip -qq v8.15.2.zip
WORKDIR /tmp/libvips-8.15.2
RUN meson setup build --libdir=lib --buildtype=release -D introspection=disabled && cd build && meson compile && meson test && meson install


WORKDIR /app
COPY --from=build /app/publish .


# ENV VIPSHOME=/usr/local/lib


ENV DOTNETCORE_ENVIRONMENT=$ENVIRONMENT
ENV SHUSAEncryption=testing
ENV PATH="$PATH:/usr/local/lib"
ENV PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
ENV LD_LIBRARY_PATH=/usr/local/lib
```

Everything builds fine and all configurations appear correct (see attached images)
<img width="1489" alt="Screenshot 2024-06-28 at 6 25 44 AM" src="https://github.com/libvips/libvips/assets/637006/389936e9-cc7d-4430-b6c6-f4d004a01c69">
<img width="687" alt="Screenshot 2024-06-28 at 6 24 00 AM" src="https://github.com/libvips/libvips/assets/637006/e11ed6f1-cfe3-4a74-92fe-4d20d70324c0">

all libraries are in place, environment variables are correct however when we run the application we still receive 

no such operation openslideload

Any thoughts ideas suggestions?

**Comment by jcupitt** [comment]
Hi, you're installing the system libvips:

```
RUN apt update && apt-get install -o Acquire::https::Verify-Peer=false -y --no-install-recommends \
	build-essential \
	cmake \
	meson \
	unzip \
...
	libvips-dev \
```

And then also building your own version from source. Just removing `libvips-dev` from the list of packages to install will probably fix it.

**Reply by jcupitt** [reply]
Oh, you also have:

```
ENV PATH="$PATH:/usr/local/lib"
```

I guess this is a typo, you need:

```
ENV PATH="$PATH:/usr/local/bin"
```

Though I think jammy might have that on the default path, I forget.

**Comment by mgbee8** [comment]
John,
Thanks for such a quick response still no luck!

so i removed libvips-dev from the install 
checked and /usr/local/bin is in the path
still get 

NetVips.VipsException: no such operation openslideload
VipsOperation: class \"openslideload\" not found

at NetVips.Operation.NewFromName(String operationName)
at NetVips.Introspect..ctor(String operationName)
at NetVips.Introspect.<>c.<Get>b__10_0(String name)
 at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
 at NetVips.Introspect.Get(String operationName)
 at NetVips.Operation.Call(String operationName, VOption kwargs, Image matchImage, Object[] args)
 at NetVips.Operation.Call(String operationName, VOption kwargs, Object[] args)
 at NetVips.Image.Openslideload(String filename, Nullable`1 level, Nullable`1 autocrop, String associated, Nullable`1 attachAssociated, Nullable`1 rgb, Nullable`1 memory, Nullable`1 access, Nullable`1 failOn, Nullable`1 revalidate)
 at shusa_digital_pathology_image_ingester.Services.ImageProcessor.GenerateDzi() in /app/shusa_digital_pathology_image_ingester/Services/ImageProcessor.cs:line 149

**Comment by kleisauke** [comment]
Did you remove the [`NetVips.Native`](https://www.nuget.org/packages/NetVips.Native) package? This NuGet package is unnecessary if you have libvips installed globally, which is also why it can be optionally installed. Keeping it may cause conflicts.

**Comment by mgbee8** [comment]
NoI didn't

**Reply by kleisauke** [reply]
I would remove it to avoid possible library conflicts. If that doesn't work, you can use `vips --vips-config` to check if libvips is built against OpenSlide.

**Comment by mgbee8** [comment]
Here is the output of vips --vips-config

enable debug: false
enable deprecated: true
enable modules: true
enable cplusplus: true
enable RAD load/save: true
enable Analyze7 load/save: true
enable PPM load/save: true
enable GIF load: true
use fftw for FFTs: false
SIMD support with highway: false
accelerate loops with ORC: true
ICC profile support with lcms: true
zlib: true
text rendering with pangocairo: false
font file support with fontconfig: false
EXIF metadata support with libexif: true
JPEG load/save with libjpeg: true
JXL load/save with libjxl: false (dynamic module: false)
JPEG2000 load/save with OpenJPEG: true
PNG load/save with libspng: false
PNG load/save with libpng: true
selected quantisation package: none
TIFF load/save with libtiff: true
image pyramid save with libarchive: false
HEIC/AVIF load/save with libheif: false (dynamic module: false)
WebP load/save with libwebp: true
PDF load with PDFium: false
PDF load with poppler-glib: false (dynamic module: false)
SVG load with librsvg: true
EXR load with OpenEXR: true
OpenSlide load: true (dynamic module: true)
Matlab load with libmatio: false
NIfTI load/save with niftiio: false
FITS load/save with cfitsio: false
GIF save with cgif: false
selected Magick package: none (dynamic module: false)
Magick API version: none
Magick load: false
Magick save: false

**Comment by mgbee8** [comment]
Different error with netVips.Native removed
NetVips.VipsException: no such operation dzsave
VipsOperation: class \"dzsave\" not found

   at NetVips.Operation.NewFromName(String operationName)
   at NetVips.Introspect..ctor(String operationName)
   at NetVips.Introspect.<>c.<Get>b__10_0(String name)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at NetVips.Introspect.Get(String operationName)
   at NetVips.Operation.Call(String operationName, VOption kwargs, Image matchImage, Object[] args)
   at NetVips.ExtensionMethods.Call(Image image, String operationName, VOption kwargs, Object[] args)
   at NetVips.Image.Dzsave(String filename, String imagename, Nullable`1 layout, String suffix, Nullable`1 overlap, Nullable`1 tileSize, Nullable`1 centre, Nullable`1 depth, Nullable`1 angle, Nullable`1 container, Nullable`1 compression, Nullable`1 regionShrink, Nullable`1 skipBlanks, String id, Nullable`1 q, Nullable`1 keep, Double[] background, Nullable`1 pageHeight, String profile)
   at shusa_digital_pathology_image_ingester.Services.ImageProcessor.GenerateDzi() in /app/shusa_digital_pathology_image_ingester/Services/ImageProcessor.cs:line 147

**Reply by kleisauke** [reply]
You'll need to install `libarchive-dev` for image pyramid save.

**Comment by jcupitt** [comment]
You've got quite a few packages you don't need, and you're missing some others. I made you a demo dockerfile:

```dockerfile
FROM ubuntu:jammy

ENV DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y \
        build-essential \
        software-properties-common \
        ninja-build \
        python3-pip \
        pkg-config \
        wget 

# we need meson for openslide and libvips build
RUN pip3 install meson

WORKDIR /usr/local/src
ENV LD_LIBRARY_PATH=/usr/local/lib

# needed for DICOM WSI in openslide
ARG DICOM_VERSION=1.1.0
ARG DICOM_URL=https://github.com/ImagingDataCommons/libdicom/releases/download/

RUN wget ${DICOM_URL}/v${DICOM_VERSION}/libdicom-${DICOM_VERSION}.tar.xz

RUN tar xfJ libdicom-${DICOM_VERSION}.tar.xz \
        && cd libdicom-${DICOM_VERSION} \
        && meson setup --libdir=lib --buildtype=release builddir \
        && meson compile -C builddir \
        && meson install -C builddir

# openslide dependencies
RUN apt-get install -y \
        glib-2.0-dev \
        libcairo2-dev \
        libgdk-pixbuf-2.0-dev \
        libjpeg-turbo8-dev \
        libopenjp2-7-dev \
        zlib1g-dev \
        libtiff-dev \
        libxml2-dev \
        libpng-dev \
        libsqlite3-dev 

ARG OPENSLIDE_VERSION=4.0.0
ARG OPENSLIDE_URL=https://github.com/openslide/openslide/releases/download

RUN wget ${OPENSLIDE_URL}/v${OPENSLIDE_VERSION}/openslide-${OPENSLIDE_VERSION}.tar.xz

RUN tar xfJ openslide-${OPENSLIDE_VERSION}.tar.xz \
        && cd openslide-${OPENSLIDE_VERSION} \
        && meson setup --libdir=lib --buildtype=release builddir \
        && meson compile -C builddir \
        && meson install -C builddir

# stuff we need to build our own libvips ... this is a pretty random selection
# of dependencies, you'll want to adjust these
RUN apt-get install -y \
        libexpat-dev \
        librsvg2-dev \
        libarchive-dev \
        libexif-dev \
        liblcms2-dev \
        libheif-dev \
        libhwy-dev

# build the head of the stable 8.15 branch
ARG VIPS_BRANCH=8.15
ARG VIPS_URL=https://github.com/libvips/libvips/tarball

RUN mkdir libvips-${VIPS_BRANCH} \
        && cd libvips-${VIPS_BRANCH} \
        && wget ${VIPS_URL}/${VIPS_BRANCH} -O - | \
                tar xfz - --strip-components 1

# "--libdir lib" makes it put the library in /usr/local/lib
# we don't need GOI
RUN cd libvips-${VIPS_BRANCH} \
        && rm -rf build \
        && meson setup build --libdir lib -Dintrospection=disabled \
        && cd build \
        && ninja \
        && ninja test \
        && ninja install

# pyvips .. handy for testing
RUN pip3 install pyvips
```

I can build and run like this:

```
$ docker build -t libvips-ubuntu22.04 .
... lots of stuff
$ docker run -it --rm -v $PWD:/data libvips-ubuntu22.04
root@ffb628171971:/usr/local/src# python3
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> pyvips.API_mode
True
>>> pyvips.Image.black(1000,1000).dzsave("x")
>>> ^D 
root@ffb628171971:/usr/local/src# 
```

**Reply by jcupitt** [reply]
(you need libdicom as well if you want to be able to load WSI DICOMs in openslide)

**Reply by jcupitt** [reply]
(I built everything in a set of subdirectories in `/usr/local/src`, I think that's neater than using `/tmp`, but it doesn't really matter)

**Comment by mgbee8** [comment]
thank you Jon. I will modify accordingly

**Comment by mgbee8** [comment]
I copied things from bits and scraps I collected. It is working as expected now thank you Both John and Kleis for the help and for the whole libvips ecosystem

**Comment by mgbee8** [comment]
Again thanks and I am closing this discussion
